language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java

```diff
@@ -0,0 +1,670 @@
+package org.ballerinalang.langserver.completions.providers.context.util;
+
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.ReBracedQuantifierNode;
+import io.ballerina.compiler.syntax.tree.ReFlagExpressionNode;
+import io.ballerina.compiler.syntax.tree.ReFlagsOnOffNode;
+import io.ballerina.compiler.syntax.tree.ReQuantifierNode;
+import io.ballerina.compiler.syntax.tree.ReUnicodeGeneralCategoryNode;
+import io.ballerina.compiler.syntax.tree.ReUnicodePropertyEscapeNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.TemplateExpressionNode;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.ballerinalang.langserver.commons.completion.LSCompletionItem;
+import org.ballerinalang.langserver.completions.SnippetCompletionItem;
+import org.ballerinalang.langserver.completions.providers.context.TemplateExpressionNodeContext.RegexTemplateNodeFinder;
+import org.ballerinalang.langserver.completions.util.ItemResolverConstants;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.ballerinalang.langserver.completions.util.SnippetBlock;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Util to generate completions for regexp.
+ *
+ * @since 2201.5.0
+ */
+public class RegexpCompletionProvider {
+
+    private static final List<String> wordSeparatorArray = Arrays.asList("`", "~", "!", "@", "#", "$", "%", "^", "&",
+            "*", "(", ")", "-", "=", "+", "[", "{", "]", "}", "\\", "|", ";", ":", "'", "\"", ",", ".", "<", ">", "/",
+            "?");
+
+    private static final List<String> reFlags = Arrays.asList("m", "i", "s", "x");
+
+    public static List<LSCompletionItem> getRegexCompletions(NonTerminalNode nodeAtCursor,
+                                                             BallerinaCompletionContext ctx,
+                                                             RegexTemplateNodeFinder nodeFinder) {
+        List<LSCompletionItem> completionItems = new ArrayList<>();
+        if (nodeAtCursor.kind() == SyntaxKind.LIST) {
+            nodeAtCursor.parent().accept(nodeFinder);
+        } else {
+            nodeAtCursor.accept(nodeFinder);
+        }
+        String resolvedWord = null;
+
+        if (nodeAtCursor.kind() == SyntaxKind.RE_QUOTE_ESCAPE
+                && nodeAtCursor.toSourceCode().equals(SyntaxKind.BACK_SLASH_TOKEN.stringValue())) {
+            // Eg: re `\<cursor>`
+            completionItems.add(new SnippetCompletionItem(ctx,
```

Comment:
Just wondering whether these completions can be pre-created completion item list. Any con to doing that?

Reply:
That's also possible. We can predeclare it in a list and when it comes to the correct context extend the completionItems with the declared list. Should I do that or not?

Reply:
Let's do that

Reply:
The only way to do this is to keep the completion list at the class level in a final static variable. But the completion items are context dependent. So we have to create completion items for each request. So what have suggested is not possible.